drivers: led: Add LED driver support for TI LP5562

JoeHut opened the pull request:
Add driver support and sample for the TI LP5562 4 channel I2C-based LED driver.

This driver supports the entire currently available API and a sample is provided.

Commit:
drivers: led: Add LED driver support for TI LP5562

Add support for TI LP5562 I2C 4-channel LED driver.

Supported blinking period: 1ms - 1000ms
Supported brightness value: 0% - 100%

This driver supports the entire currently available API.

Signed-off-by: Johannes Hutter <johannes@proglove.de>

Commit:
samples: drivers: Add sample application for LP5562

Add a sample application to show the LP5562 driver. The three RGB
channels are used to set multiple colors after each other under the
assumption that accordingly colored LEDs are connected. Blinking is also
shown in multiple colors.

Signed-off-by: Johannes Hutter <johannes@proglove.de>

galak reviewed (commented):
Device tree support for this?

JoeHut commented:
@galak I don't really get how the dts works in that case... I just looked at the PCA9633 sample. How do you determine at what register the device is added to the I2C node?

JoeHut commented:
@galak I updated it for device tree compatibility but I am not sure whether it is done the right way. As I understand it, we will never get rid of fixup files for I2C slaves because the architecture specific definitions generated from the dts have to be mapped. Is this correct? Or am I missing the clean approach?

Mani-Sadhasivam commented:
@galak Not sure why I'm not being notified for the LED API/Driver Pull Requests even though my github handle is being listed in CODEOWNERS!

Mani-Sadhasivam commented:
@JoeHut Please give me some time to look into this. I have just seen this PR and the absence of prompt notification added the delay.

Thanks!

JoeHut commented:
@Mani-Sadhasivam No problem. If you are the LED codeowner we can maybe talk about RGB and brightness fading extensions of the API afterwards.

Mani-Sadhasivam commented:
@JoeHut Sure. We can do the API discussion in #7619

dbkinder commented on samples/drivers/led_lp5562/README.rst:
```diff
@@ -0,0 +1,47 @@
+For flashing the application, refer to the Flashing section of the
+:ref:`nrf52840_pca10056` board documentation.
+
+References
```
I'd recommend deleting this heading and change the bullet to this and change the reference noted above:

    .. _LP5562 Manual: http://www.ti.com/lit/ds/symlink/lp5562.pdf

JoeHut replied:
done. Thanks for the feedback.

dbkinder commented on samples/drivers/led_lp5562/README.rst:
```diff
@@ -0,0 +1,47 @@
+ 12. blink the LEDs in purple
+ 13. turn off LEDs
+
+The sample assumes that RGB LEDs are connected in the intended colors to the
```
Replace these two sentences with:

    Refer to the `LP5562 Manual`_ for the RGB LED connections and color
    channel mappings used by this sample.

Mani-Sadhasivam commented on drivers/led/Kconfig.lp5562:
```diff
@@ -0,0 +1,42 @@
+
+config LP5562_I2C_ADDRESS
+	hex "LP5562 I2C slave address"
+	range 0x30 0x67
```
range should be from `0x30` to `0x33`

JoeHut replied:
ups. That's a copy and paste fail... Thanks

Mani-Sadhasivam commented on drivers/led/lp5562.c:
```diff
@@ -0,0 +1,837 @@
+ * @param channel LED channel.
+ * @param source Buffer for the register.
+ *
+ * @return 0 if successful, -ERRNO otherwise.
```
You can directly say that we are returning `-EINVAL` for error cases instead of `-ERRNO`

JoeHut replied:
I change the return docs to use ``@retval``

Mani-Sadhasivam commented on drivers/led/lp5562.c:
```diff
@@ -0,0 +1,837 @@
+ *	brightness value instead of using the execution engines.
+ *
+ * @param channel LED channel.
+ * @param source Buffer for the register.
```
How about
@param reg Pointer to the register address

JoeHut replied:
ack

Mani-Sadhasivam commented on drivers/led/lp5562.c:
```diff
@@ -0,0 +1,837 @@
+ * @return 0 if successful, -ERRNO otherwise.
+ */
+static int lp5562_get_engine_ram_base_addr(enum lp5562_led_sources engine,
+					   u8_t *base_addr)
```
Please fix up the alignment here

JoeHut replied:
Can you check in an editor? In vim it looks like the alignment is right, so maybe it is just the GitHub rendering.

Mani-Sadhasivam replied:
I was referrig to the mis alignment regarding the second parameter `u8_t *base_addr`. It should have started at the level after opening parentheses (i.e) below `e`. Btw, this is just a nitpick which will make your code look good :)

JoeHut replied:
sure, that's what I understood and I actually did align those arguments. What I see locally is different to what I see in the GitHub rendering :man_shrugging: 
 
![lp5562](https://user-images.githubusercontent.com/18048777/43153736-67e6a4b2-8f72-11e8-9089-d83a55f2e27c.jpg)

But then again I am relying on a 8 space width of tabs for that of course. But this is what the other drivers look like as well

Mani-Sadhasivam commented on drivers/led/lp5562.c:
```diff
@@ -0,0 +1,837 @@
+ * @param engine Engine the base address is requested for.
+ * @param base_addr Buffer for the base address.
+ *
+ * @return 0 if successful, -ERRNO otherwise.
```
Again, please mention the actual error value

Mani-Sadhasivam commented on drivers/led/lp5562.c:
```diff
@@ -0,0 +1,837 @@
+ * @brief Get the base address for programs of the given execution engine.
+ *
+ * @param engine Engine the base address is requested for.
+ * @param base_addr Buffer for the base address.
```
Please change it as per my previous comment

Mani-Sadhasivam commented on drivers/led/lp5562.c:
```diff
@@ -0,0 +1,837 @@
+ * @param engine Engine the shift is requested for.
+ * @param shift Buffer for the shift value.
+ *
+ * @param 0 if successful, -ERRNO otherwise.
```
Ditto

Mani-Sadhasivam commented on drivers/led/lp5562.c:
```diff
@@ -0,0 +1,837 @@
+ * OP_MODE and ENABLE registers.
+ *
+ * @param engine Engine the shift is requested for.
+ * @param shift Buffer for the shift value.
```
Ditto

Mani-Sadhasivam commented on drivers/led/lp5562.c:
```diff
@@ -0,0 +1,837 @@
+static int lp5562_ms_to_prescale_and_step(struct led_data *data, u32_t ms,
+					  u8_t *prescale, u8_t *step_time)
+{
+	if ((ms < data->min_period) || (ms > data->max_period)) {
```
Check here is superfluous

JoeHut replied:
hm, I am no fan of allowing invalid arguments, but I can remove the check and add a comment.

Mani-Sadhasivam replied:
What I meant was, the argument check is already done in the caller `lp5562_program_ramp`, so you don't need to do the same in this function also. That's why I mentioned it as superfluous!

Mani-Sadhasivam commented on drivers/led/lp5562.c:
```diff
@@ -0,0 +1,837 @@
+	}
+
+	/*
+	 * step_time with prescaler as 0 is 0.49ms. max value for step_time is
```
Nit: Please start with capital letter for comments

Mani-Sadhasivam commented on drivers/led/lp5562.c:
```diff
@@ -0,0 +1,837 @@
+	 */
+	*prescale = 1;
+	*step_time = ms >> 4;
+	return 0;
```
Insert new line before return

Mani-Sadhasivam commented on drivers/led/lp5562.c:
```diff
@@ -0,0 +1,837 @@
+static int lp5562_get_available_engine(struct device *dev,
+				       enum lp5562_led_sources *engine)
+{
+	enum lp5562_led_sources i;
```
How about
`enum lp5562_led_sources src;`

Mani-Sadhasivam commented on drivers/led/lp5562.c:
```diff
@@ -0,0 +1,837 @@
+ * @param engine Engine the operational mode is changed for.
+ * @param mode Mode to set.
+ *
+ * @return 0 if successful, -ERRNO otherwise.
```
Ditto

Mani-Sadhasivam commented on drivers/led/lp5562.c:
```diff
@@ -0,0 +1,837 @@
+ * @param engine Engine the execution state is changed for.
+ * @param state State to set.
+ *
+ * @return 0 if successful, -ERRNO otherwise.
```
Ditto

Mani-Sadhasivam commented on drivers/led/lp5562.c:
```diff
@@ -0,0 +1,837 @@
+ * @param dev LP5562 device.
+ * @param engine Engine that is started.
+ *
+ * @return 0 if successful, -ERRNO otherwise.
```
Please change all below occurences

Mani-Sadhasivam reviewed (changes requested):
Diver looks good overall. Just mentioned several nitpicks. Since this driver's working mode is complex, it'd be good to add a description of how the driver operates as a brief comment at the start of `lp5562.c`

Mani-Sadhasivam commented:
@JoeHut You still need to add a devicetree node for this driver in a board which utilizes it!

JoeHut commented:
@Mani-Sadhasivam I addressed your requests. I added a device tree overlay in the sample I provided for the nrf52840_pca10056 board I used for development. I don't know of a standard/devkit board that uses this driver. I implemented it for our own custom board.

dbkinder reviewed (approved):
+1 for docs, thanks!

JoeHut commented:
@Mani-Sadhasivam Could you re-review at some point, please :)

Mani-Sadhasivam commented on drivers/led/lp5562.c:
```diff
@@ -0,0 +1,879 @@
+			/*
+			 * API expectes set_brightness to have no effect if the
+			 * LED is blinking already.
+			 */
```
This comment is not true! API states that,`set_brightness` should have no effect on the blinking rate if called after `led_blink`.

JoeHut replied:
ouh, than I misinterpreted that:
``Calling this function after led_blink() won't affect blinking.``
I thought this means the blinking is not changed at all. So what should happen is that the LED keeps blinking with the previously defined rate but the brightness should change?

Mani-Sadhasivam replied:
yes, only the brightness is expected to change but the LED should continue blinking at the same rate.

Mani-Sadhasivam reviewed (changes requested):
Hi @JoeHut,

Driver and Sample application looks good. Added one inline comment regarding the LED API. Other than that, you also need to fix up the commits. Sample application commit should not have driver changes.

JoeHut commented:
@Mani-Sadhasivam I fixed up the commits (sorry for that). I changed the ``set_brightness`` implementation to keep the blinking rate but change the brightness, if ``led_blink`` was called before. It is a little bit hacky, but I think the API was designed with a certain functionality in mind that not all LED drivers provide necessarily. I think it might make sense to also change the comment in the LED API in order to make sure, that the brightness should be changed, but the blinking rate not. At least for me it was not clear :)

Mani-Sadhasivam commented:
LGTM! @JoeHut Thanks a lot for contributing the driver :-) I know that we are very much limited with the API for these kind of complex devices, but obviously we can't do much here. One suggestion I would say it to have an API which does chip specific configuration (like programming the engine in this case). This was pointed out by @iha during the initial LED PR and still yet to be implemented.

JoeHut commented:
Thanks for the feedback and review, @Mani-Sadhasivam! I think the problem is that the range of devices to support with this API is quite large with LEDs that can just be turned on or off via GPIO to proper multi-color chips with fading and blinking. Do you think adding brightness fading does not make sense to add to the API? In the initial PR @mbolivar planned to add RGB support, but I don't know if he has time for that.

If we won't change the general LED API, should I implement those things rather with an API exposed in ``include/drivers/led``? I think that would make more sense than passing some void pointers.

Mani-Sadhasivam commented:
@JoeHut I didn't say that the API cannot be extended ;-) It should be extended to cover wide range of LED devices (like this one). Still I haven't looked at your API proposal but if you have any idea to extend the API to support more devices please submit a PR :)

RGB support is something which is missing from day one and I don't have any exact time frame to add that. Regarding the brightness fading, yeah we can add that to API (sceptical) if we also have a driver utilising that ;-)
